src/core/thread/anycast_locator.cpp

```diff
@@ -0,0 +1,163 @@
+    , mCallback(nullptr)
+    , mContext(nullptr)
+    , mAnycastLocate(UriPath::kAnycastLocate, HandleAnycastLocate, this)
+
+{
+    Get<Tmf::Agent>().AddResource(mAnycastLocate);
+}
+
+Error AnycastLocator::Locate(const Ip6::Address &aAnycastAddress, Callback aCallback, void *aContext)
+{
+    Error            error   = kErrorNone;
+    Coap::Message *  message = nullptr;
+    Ip6::MessageInfo messageInfo;
+
+    VerifyOrExit((aCallback != nullptr) && Get<Mle::Mle>().IsAnycastLocator(aAnycastAddress),
+                 error = kErrorInvalidArgs);
+
+    if (mCallback != nullptr)
+    {
+        // We set the `mCallback` to `nullptr` first and then abort the
+        // ongoing CoAP transaction (if any). Note that the abort method
+        // call can invoke the response callback `HandleResponse()` from
+        // which we invoke the `mCallback`.
+
+        mCallback = nullptr;
+        IgnoreError(Get<Tmf::Agent>().AbortTransaction(HandleResponse, this));
+    }
+
+    message = Get<Tmf::Agent>().NewMessage();
+    VerifyOrExit(message != nullptr, error = kErrorNoBufs);
+
+    SuccessOrExit(error = message->InitAsConfirmablePost(UriPath::kAnycastLocate));
+    SuccessOrExit(error = message->SetPayloadMarker());
+
+    // No payload, so we need to remove CoAP payload marker.
+    IgnoreError(message->SetLength(message->GetLength() - 1));
+
+    messageInfo.SetSockAddr(Get<Mle::MleRouter>().GetMeshLocal16());
+    messageInfo.SetPeerAddr(aAnycastAddress);
+    messageInfo.SetPeerPort(Tmf::kUdpPort);
+
+    SuccessOrExit(error = Get<Tmf::Agent>().SendMessage(*message, messageInfo, HandleResponse, this));
+
+    mCallback = aCallback;
+    mContext  = aContext;
+
+exit:
+    FreeMessageOnError(message, error);
+    return error;
+}
+
+void AnycastLocator::HandleResponse(void *               aContext,
+                                    otMessage *          aMessage,
+                                    const otMessageInfo *aMessageInfo,
+                                    Error                aResult)
+{
+    static_cast<AnycastLocator *>(aContext)->HandleResponse(
+        static_cast<Coap::Message *>(aMessage), static_cast<const Ip6::MessageInfo *>(aMessageInfo), aResult);
+}
+
+void AnycastLocator::HandleResponse(Coap::Message *aMessage, const Ip6::MessageInfo *aMessageInfo, Error aResult)
+{
+    OT_UNUSED_VARIABLE(aMessageInfo);
+
+    uint16_t            rloc16  = Mac::kShortAddrInvalid;
+    const Ip6::Address *address = nullptr;
+    Ip6::Address        meshLocalAddress;
+
+    SuccessOrExit(aResult);
+    OT_ASSERT(aMessage != nullptr);
+
+    meshLocalAddress.SetPrefix(Get<Mle::Mle>().GetMeshLocalPrefix());
+    SuccessOrExit(Tlv::Find<ThreadMeshLocalEidTlv>(*aMessage, meshLocalAddress.GetIid()));
+    SuccessOrExit(Tlv::Find<ThreadRloc16Tlv>(*aMessage, rloc16));
+
+    address = &meshLocalAddress;
+
+exit:
+    if (mCallback != nullptr)
+    {
+        mCallback(mContext, address, rloc16);
+        mCallback = nullptr;
```

Comment:
```suggestion
        mCallback = nullptr;
        mCallback(mContext, address, rloc16);
```

Set `mCallback` to null first? User may query again in `mCallback`.